File: selfcsgame/selfcsgame/trrne/lottery.cs
namespace trrne;

public class LotteryPair<T>
{
    readonly List<T> subjects = [];
    readonly List<double> weights = [];

    public LotteryPair(params (T subject, double weight)[] pairs)
    {
        Len = pairs.Length;
        Array.ForEach(pairs, pair =>
        {
            subjects.Add(pair.subject);
            weights.Add(pair.weight);
        });
    }

    public int Len { get; private set; }

    public T[] Subjects => [.. subjects];
    public double[] Weights => [.. weights];

    public double TotalWeights => weights.Sum();

    public object this[int index] => (subjects[index], weights[index]);

    public T Weighted() => Subjects[Lottery.Bst(Weights)];
}

public static class Lottery
{
    readonly static Random rand;
    static Lottery() => rand = new();

    public static int Bst(params double[] weights)
    {
        if (weights.Length <= 0)
        {
            return -1;
        }

        double[] totals = new double[weights.Length];
        double total = 0;
        for (int i = 0; i < weights.Length; i++)
        {
            total += weights[i];
            totals[i] = total;
        }
        double random = rand.NextDouble() * total;
        int bottom = 0, top = totals.Length - 1;
        while (bottom < top)
        {
            int middle = (bottom + top) / 2;
            if (random > totals[middle])
            {
                bottom = middle + 1;
            }
            else
            {
                if (random >= (middle > 0 ? totals[middle - 1] : 0))
                {
                    return middle;
                }
                top = middle;
            }
        }
        return top;
    }
}

File: selfpygame/src/lottery.py
from typing import Generic, TypeVar
from random import uniform

T = TypeVar('T')


class LotteryPair(Generic[T]):
    def __init__(self, pairs: list[tuple[T, float]]) -> None:
        super().__init__()
        self.__pairs: list[tuple[T, float]] = pairs
        self.__subjects: list[T] = []
        self.__weights: list[float] = []
        for i in range(len(pairs)):
            self.__subjects.append(pairs[i][0])
            self.__weights.append(pairs[i][1])

    def length(self) -> int: return len(self.__pairs)

    def pairs(self) -> list[tuple[T, float]]: return self.__pairs
    def subjects(self) -> list[T]: return self.__subjects
    def weights(self) -> list[float]: return self.__weights

    def total_weight(self) -> float:
        dst = 0
        for weight in self.__weights:
            dst += weight
        return dst


class Lottery(Generic[T]):
    @staticmethod
    def bst(weights: list[float]) -> int:
        '''二分探索木'''
        if len(weights) <= 0:
            # raise Exception('からっぽやんけ')
            return -1

        totals: list[float] = []
        total: float = 0.0
        for i in range(len(weights)):
            total += weights[i]
            totals.append(total)
        # for i in totals:
        #     print(f'totals: {i}')

        r: float = uniform(0, total)
        bottom: int = 0
        top: int = len(totals)-1
        while bottom < top:
            center: int = int((bottom+top)/2)
            if r > totals[center]:
                bottom = center+1
            else:
                if r >= totals[center-1] if center > 0 else 0.0:
                    return center
                top = center
        return top

    @staticmethod
    def weighted(pairs: LotteryPair[T]) -> T:
        if pairs.length() <= 0:
            raise Exception()
        return pairs.__subjects[Lottery.bst(pairs.__weights)] if pairs.length() != 1 else pairs.__subjects[0]
